File: day1/part2.cpp
#include <string>
#include <iostream>

int main() {
    // CAPTCHA string
	  std::string captcha = "12131415";
    
    // Sum of all digits that match next digit in the list
    int sum = 0;
    
    // Looking at the digit halfway around the list (captcha has even number of digits)
    int step = captcha.length() / 2;
    
    // Loop through all digits and see if they match the one halfway around the list
    for (unsigned int i = 0; i < captcha.length(); i++)
    {
        // Modulo to wrap-around if i+step goes past the end of the captcha string
        if (captcha[i] == captcha[(i+step)%captcha.length()])
        {
            // Increment the sum by the digit value (converted from char by subtracting ascii value of '0')
            sum += (captcha[i] - '0');
        }
    }
    
    // Print the result
    printf ("%d\n", sum);
}

File: day1/part2.rs
fn main() {
    let captcha = "12131415";
    
    // Turn the string into an array of bytes, and add indices
    let captcha: Vec<u8> = captcha.bytes().collect();
    let digits = captcha.iter().enumerate();
    
    // Keep digit if the digit halfway around the (circular) array matches
    let digits = digits.filter(|&(i, e)| captcha[(i+captcha.len()/2)%captcha.len()] == *e);
    
    // Convert from ascii digit to value of the digit (u32 to avoid overflow final sum)
    let digits = digits.map(|(_, e)| (e-b'0') as u32);
    
    // Get the result
    let digits = digits.collect::<Vec<_>>();
    let sum: u32 = digits.iter().sum();
    println!("{}", sum);
}
